core/services/mongodb-remote/src/test/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizerUnitTests.kt

```diff
@@ -79,60 +96,849 @@ class DataSynchronizerUnitTests {
+                StitchServiceException("E11000", StitchServiceErrorCode.MONGODB_ERROR)
+            )
 
-        // unfreeze the configs and reload
-        dataSynchronizer.getSynchronizedDocuments(namespace).forEach {
-            it.isFrozen = false
+            dataSynchronizer.insertOneAndSync(namespace, document2)
+
+            assertEquals(
+                document2,
+                withoutVersionId(
+                    dataSynchronizer.findOneById(
+                        namespace,
+                        document2["_id"],
+                        BsonDocument::class.java,
+                        CodecRegistries.fromCodecs(BsonDocumentCodec()))))
+
+            dataSynchronizer.doSyncPass()
+
+            emitEventSemaphore.acquire()
+
+            verify(changeEventListener, times(2)).onEvent(eq(document2["_id"]), any())
+            verify(conflictHandler, times(1)).resolveConflict(eq(document2["_id"]), any(), any())
+            verify(errorListener, times(0)).onError(eq(document2["_id"]), any())
+
+            assertNull(
+                dataSynchronizer.findOneById(namespace, document2["_id"], BsonDocument::class.java, CodecRegistries.fromCodecs(BsonDocumentCodec()))
+            )
         }
-        dataSynchronizer.reloadConfig()
+    }
 
-        // this time ensure that the remote insert has been called
-        dataSynchronizer.doSyncPass()
+    @Test
+    fun testFailedInsert() {
+        withNewDataSynchronizer { dataSynchronizer, coreRemoteMongoCollectionMock ->
+            val document2 = newDoc()
+            val emitEventSemaphore = Semaphore(0)
+            val emitErrorSemaphore = Semaphore(0)
+
+            val expectedLocalEvent = ChangeEvent.changeEventForLocalInsert(namespace, document2, true)
+            val expectedRemoteEvent = ChangeEvent.changeEventForLocalDelete(namespace, document2["_id"], false)
+
+            val changeEventListener = changeEventListener(emitEventSemaphore, expectedRemoteEvent)
+            val conflictHandler = conflictHandler(expectedLocalEvent, expectedRemoteEvent)
+            val errorListener = errorListener(emitErrorSemaphore, document2["_id"]!!)
+
+            dataSynchronizer.configure(namespace,
+                conflictHandler,
+                changeEventListener,
+                errorListener,
+                BsonDocumentCodec())
+
+            // force a conflict
+            `when`(coreRemoteMongoCollectionMock.insertOne(any())).thenThrow(StitchServiceException("bad", StitchServiceErrorCode.UNKNOWN))
+
+            dataSynchronizer.insertOneAndSync(namespace, document2)
+
+            dataSynchronizer.doSyncPass()
+
+            emitErrorSemaphore.acquire()
+
+            verify(changeEventListener, times(1)).onEvent(eq(document2["_id"]), any())
+            verify(conflictHandler, times(0)).resolveConflict(eq(document2["_id"]), any(), any())
+            verify(errorListener, times(1)).onError(eq(document2["_id"]), any())
+
+            assertEquals(
```

Comment:
[minor] I think in general, for all these tests, we should add a comment when checking the document in the local collection to explain why we're checking for the specific expected state. I was confused at first why we were expecting the document to exist when the insert failed, but I realized it's because we expect the document to be there locally, just not remotely.